dbms/src/DataStreams/FilterTransformAction.cpp

```diff
@@ -0,0 +1,181 @@
+} // namespace ErrorCodes
+
+
+FilterTransformAction::FilterTransformAction(
+    const Block & header_,
+    const ExpressionActionsPtr & expression_,
+    const String & filter_column_name)
+    : header(header_)
+    , expression(expression_)
+{
+    /// Determine position of filter column.
+    expression->execute(header);
+
+    filter_column = header.getPositionByName(filter_column_name);
+    auto & column_elem = header.safeGetByPosition(filter_column);
+
+    /// Isn't the filter already constant?
+    if (column_elem.column)
+        constant_filter_description = ConstantFilterDescription(*column_elem.column);
+
+    if (!constant_filter_description.always_false && !constant_filter_description.always_true)
+    {
+        /// Replace the filter column to a constant with value 1.
+        FilterDescription filter_description_check(*column_elem.column);
+        column_elem.column = column_elem.type->createColumnConst(header.rows(), UInt64(1));
+    }
+}
+
+bool FilterTransformAction::alwaysFalse()
+{
+    return constant_filter_description.always_false;
+}
+
+Block FilterTransformAction::getHeader() const
+{
+    return header;
+}
+
+ExpressionActionsPtr FilterTransformAction::getExperssion() const
+{
+    return expression;
+}
+
+bool FilterTransformAction::transform(Block & block)
+{
+    if (unlikely(!block))
+        return true;
+
+    expression->execute(block);
+
+    if (constant_filter_description.always_true)
+        return true;
+
+    size_t columns = block.columns();
+    size_t rows = block.rows();
+    ColumnPtr column_of_filter = block.safeGetByPosition(filter_column).column;
+
+    /** It happens that at the stage of analysis of expressions (in sample_block) the columns-constants have not been calculated yet,
+        *  and now - are calculated. That is, not all cases are covered by the code above.
+        * This happens if the function returns a constant for a non-constant argument.
+        * For example, `ignore` function.
+        */
+    constant_filter_description = ConstantFilterDescription(*column_of_filter);
+
+    if (constant_filter_description.always_false)
+    {
+        block.clear();
+        return true;
+    }
+
+    IColumn::Filter * filter;
+    ColumnPtr filter_holder;
+
+    if (constant_filter_description.always_true)
+    {
+        return true;
+    }
+    else
+    {
+        FilterDescription filter_and_holder(*column_of_filter);
+        filter = const_cast<IColumn::Filter *>(filter_and_holder.data);
+        filter_holder = filter_and_holder.data_holder;
+    }
+
+    /** Let's find out how many rows will be in result.
+      * To do this, we filter out the first non-constant column
+      *  or calculate number of set bytes in the filter.
+      */
+    size_t first_non_constant_column = 0;
+    for (size_t i = 0; i < columns; ++i)
+    {
+        if (!block.safeGetByPosition(i).column->isColumnConst())
+        {
+            first_non_constant_column = i;
+
```

Comment:
Why not add a break？

Reply:
Are you talking about the breaker on L214?